src/main/java/hu/oe/nik/szfmv/automatedcar/visualization/CourseDisplay.java

```diff
@@ -112,19 +115,35 @@ private AutomatedCar getCarObject(List<WorldObject> objects) {
      * @param g2d Buffered image.
      * @param world World object which will be drawn.
      */
-    private void drawObjects(Graphics2D g2d, World world) {
-        int[] offsets = getCarOffset(getCarObject(world.getWorldObjects()));
+    private void drawObjects(Graphics2D g2d, WorldManager world) {
+        int[] offsets = getCarOffset(world.getAutomatedCar());
 
         DebugViewer viewer = new DebugViewer(g2d);
-
-        for (WorldObject object : world.getWorldObjects()) {
+        AutomatedCar car = world.getAutomatedCar();
+        for (IObject object : world.getAllObjectsInRectangle(
+                new Position(0,0),
+                new Position(this.world.getWidth(),this.world.getHeight()))) {
+            Point2D refPoint;
+            try {
+                refPoint = new Point(object.getReferenceX(), object.getReferenceY());
+            } catch (NullPointerException e){
+                refPoint = new Point(0,0);
+            }
             AffineTransform t = new AffineTransform();
-            t.translate(object.getX() + offsets[0], object.getY() + offsets[1]);
+            t.translate(object.getPosX() - refPoint.getX() + offsets[0], object.getPosY() - refPoint.getY()+ offsets[1]);
+            t.rotate(Math.toRadians(-object.getRotation()),refPoint.getX(),refPoint.getY());
```

Comment:
here you need to add the offsets[x] where x is 0,1 to shift the refpoints correctly and in sync with the object